near/src/runtime.rs

```diff
@@ -269,15 +268,40 @@ impl RuntimeAdapter for NightshadeRuntime {
             let (epoch_hash, offset) = vm.get_epoch_offset(*prev_block_hash, block_index)?;
             if offset == 0 && epoch_hash != CryptoHash::default() {
                 vm.finalize_epoch(&epoch_hash, prev_block_hash, block_hash)?;
+                let prev_epoch_hash = vm.get_prev_epoch_hash(&epoch_hash)?;
+                let prev_prev_stake_change =
+                    vm.get_validators(prev_epoch_hash)?.stake_change.clone();
                 let prev_stake_change = vm.get_validators(epoch_hash)?.stake_change.clone();
-                for (account_id, new_stake) in vm.get_validators(*block_hash)?.stake_change.iter() {
+                println!("prev_stake_change: {:?}", prev_stake_change);
+                let stake_change = &vm.get_validators(*block_hash)?.stake_change;
+                for (account_id, new_stake) in prev_stake_change.iter() {
```

Comment:
This code is most likely wrong (it appears to completely ignore `prev_prev_stake_change`, so likely returns stake too early).

Consider converting stake_change and prev_state_change keys into btree sets, computing their union, and iterating over such union below instead of `for (account_id, new_stake) in stake_change.iter() {`, this seems to address the particular concern you are attempting to solve here.

Reply:
(discussed offline, it actually is correct since `stake_change` not having a key implies `prev_prev_stake_change[key]=0`, but still do the approach with the union).